src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,588 @@
+		 * does not referencing and does not referenced by any table
+		 */
+		MemoryContextSwitchTo(oldContext);
+		return NIL;
+	}
+	else
+	{
+		/*
+		 * Create the returning list according to the direction of reference. We
+		 * utilize indexes of each relation node (which has relation id as an
+		 * attribute) to get result from transitivity matrix easily.
+		 */
+		relationIndex = relationNode->index;
+		if (isAffecting)
+		{
+			uint32 tableCounter = 0;
+
+			while (tableCounter < frelGraph->nodeCount)
+			{
+				if (frelGraph->transitivityMatrix[relationIndex][tableCounter])
+				{
+					Oid referredTableOid = frelGraph->indexToOidArray[tableCounter];
+					foreignKeyList = lappend_oid(foreignKeyList, referredTableOid);
+				}
+
+				tableCounter += 1;
+			}
+		}
+		else
+		{
+			uint32 tableCounter = 0;
+
+			while (tableCounter < frelGraph->nodeCount)
+			{
+				if (frelGraph->transitivityMatrix[tableCounter][relationIndex])
+				{
+					Oid referringTableOid = frelGraph->indexToOidArray[tableCounter];
+					foreignKeyList = lappend_oid(foreignKeyList, referringTableOid);
+				}
+
+				tableCounter += 1;
+			}
+		}
+	}
+
+	MemoryContextSwitchTo(oldContext);
+
+	return foreignKeyList;
+}
+
+
+/*
+ * CreateForeignKeyRelationGraph creates the foreign key relation graph using
+ * foreign constraint provided by pg_constraint metadata table.
+ */
+static void
+CreateForeignKeyRelationGraph()
+{
+	SysScanDesc fkeyScan;
+	HeapTuple tuple;
+	HASHCTL info;
+	uint32 hashFlags = 0;
+	Relation fkeyRel;
+	int curIndex = 0;
+
+	/* if we have already created the graph, use it */
+	if (IsForeignKeyGraphValid())
+	{
+		return;
+	}
+	else
+	{
+		ClearForeignKeyRelationGraph();
+		frelGraph = (FRelGraph *) palloc(sizeof(FRelGraph));
+		frelGraph->validGraph = false;
+
+		/* create (oid) -> [FRelNode] hash */
+		memset(&info, 0, sizeof(info));
+		info.keysize = sizeof(Oid);
+		info.entrysize = sizeof(FRelNode);
+		info.hash = oid_hash;
+		info.hcxt = CurrentMemoryContext;
+		hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
+
+		frelGraph->nodeMap = hash_create("foreign key relation map (oid)",
+										 64 * 32, &info, hashFlags);
+	}
+
+	fkeyRel = heap_open(ConstraintRelationId, AccessShareLock);
+	fkeyScan = systable_beginscan(fkeyRel, ConstraintRelidIndexId, true,
+								  NULL, 0, NULL);
+
+	while (HeapTupleIsValid(tuple = systable_getnext(fkeyScan)))
+	{
+		Form_pg_constraint con = (Form_pg_constraint) GETSTRUCT(tuple);
+		bool referringFound = false;
+		bool referredFound = false;
+		FRelNode *referringNode = NULL;
+		FRelNode *referredNode = NULL;
+		Oid referringRelationOid = InvalidOid;
+		Oid referredRelationOid = InvalidOid;
+
+		/* Not a foreign key */
+		if (con->contype != CONSTRAINT_FOREIGN)
+		{
+			continue;
+		}
+
+		/*
+		 * Check whether we have seen  referring or referred relations in
+		 * constraint table before. If we haven't, add them to the hash map
+		 * of graph.
+		 */
+		referringRelationOid = con->conrelid;
```

Comment:
OK, so here's what the simplification could look like. The gist is that we're iterating over an _ordered_ list of `<referencingOid, referencedOid>`, so if we see the same pair in a row, we skip it. This removes any need to check whether we've already added an edge.

In addition, we don't need a node counter or current index variable because we can use the number of entries in the hash for that.

Finally, since we only process each edge once, and in order, we can blindly add them to the adjacency list for a referencing node. The code's just a sketch but I think it's easier than the four-way `if` statement you have in the original.

```c
static void
PopulateAdjacencyLists()
{
	ScanKeyData scanKey[3];
	Oid prevReferencingOid = InvalidOid;
	Oid prevReferencedOid = InvalidOid;

	/*
	 * FIXME: I haven't tested this to figure out how to make SK_ORDER_BY work,
	 * but it should, somehow
	 */
	/* first we'll order by relid */
	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid,
				BTEqualStrategyNumber, F_OIDEQ, NULL);
	scanKey[0].sk_flags |= SK_ORDER_BY;

	/* next by frelid */
	ScanKeyInit(&scanKey[1], Anum_pg_constraint_confrelid,
				BTEqualStrategyNumber, F_OIDEQ, NULL);
	scanKey[1].sk_flags |= SK_ORDER_BY;

	/* finally we only want foreign keys */
	ScanKeyInit(&scanKey[2], Anum_pg_constraint_contype,
				BTEqualStrategyNumber, F_CHAREQ,
				CharGetDatum(CONSTRAINT_FOREIGN));

	fkeyRel = heap_open(ConstraintRelationId, AccessShareLock);
	fkeyScan = systable_beginscan(fkeyRel, ConstraintRelidIndexId, true,
								  NULL, 3, scanKey);

	while (HeapTupleIsValid(tuple = systable_getnext(fkeyScan)))
	{
		Form_pg_constraint con->conrelid = (Form_pg_constraint) GETSTRUCT(tuple);

		/* we just saw this edge, no need to add it twice */
		if (con->conrelid == prevReferencingOid &&
			con->confrelid == prevReferencedOid)
		{
			continue;
		}

		AddEdge(frelGraph->nodeMap, con->conrelid, con->confrelid);

		prevReferencingOid = con->conrelid;
		prevReferencedOid = con->confrelid;
	}

	systable_endscan(fkeyScan);
	heap_close(fkeyRel, AccessShareLock);
}

static void
AddEdge(HTAB *adjacencyLists, Oid referencingOid, Oid referencedOid)
{
	FRelNode *referencing = CreateOrFindNode(adjacencyLists, referencingOid);
	FRelNode *referenced = CreateOrFindNode(adjacencyLists, referencedOid);

	referencing->adjacencyLists = lappend(referencing->adjacencyLists, referenced);
}

static FRelNode *
CreateOrFindNode(HTAB *adjacencyLists, Oid relid)
{
	bool found = false;
	Index newNodeIdx = hash_get_num_entries(adjacencyLists);
	FRelNode *node = (FRelNode *) hash_search(adjacencyLists,
	                                          relid,
											  HASH_ENTER,
											  &found);

	if (!found)
	{
		node->adjacencyList = NIL;
		node->index = newNodeIdx;
	}
}
```